bin/propolis-utils/src/bin/cpuid-gen.rs

```diff
@@ -261,15 +262,125 @@ fn print_toml(results: &BTreeMap<CpuidKey, Cpuid>) {
+                        eprintln!(
+                            "vendor in leaf 0 values ({values:?}) \
+                        not recognized, selecting default vendor"
+                        );
+                        CpuidVendor::Amd
+                    }
+                    Ok(v) => v,
+                }
+            }
+        }
+    };
+
+    // Returns `true` if `key` has subleaf data and its leaf index is `leaf`.
+    fn key_matches_leaf_and_has_subleaves(key: &CpuidKey, leaf: u32) -> bool {
+        let CpuidKey::SubLeaf(l, _) = key else {
+            return false;
+        };
+
+        leaf == *l
+    }
+
+    let cpuid = Cpuid {
+        entries: results
+            .iter()
+            .filter_map(|(key, value)| {
+                // propolis-server will complain if it receives a CPUID entry
+                // with no subleaf and a CPUID entry with a subleaf for the same
+                // leaf. Avoid emitting these by filtering out subleaf-less
+                // entries for any leaf that has an entry with subleaf data.
+                let (leaf, subleaf) = match key {
+                    CpuidKey::Leaf(leaf) => {
+                        if results.keys().any(|k| {
+                            key_matches_leaf_and_has_subleaves(k, *leaf)
+                        }) {
+                            return None;
+                        }
+
+                        (*leaf, None)
+                    }
+                    CpuidKey::SubLeaf(leaf, subleaf) => (*leaf, Some(*subleaf)),
+                };
+
+                Some(CpuidEntry {
+                    leaf,
+                    subleaf,
+                    eax: value.eax,
+                    ebx: value.ebx,
+                    ecx: value.ecx,
+                    edx: value.edx,
+                })
+            })
+            .collect(),
+        vendor,
+    };
+
+    println!("{}", serde_json::to_string_pretty(&cpuid).unwrap());
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+enum OutputFormat {
+    /// Print a human-readable plain-text representation.
+    #[default]
+    Text,
+
+    /// Print TOML suitable for use in a propolis-standalone config file.
+    Toml,
+
+    /// Print JSON suitable for inclusion in a propolis-server instance spec.
+    Json,
+}
+
+impl FromStr for OutputFormat {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "text" => Self::Text,
+            "toml" => Self::Toml,
+            "json" => Self::Json,
+            _ => {
+                return Err(
+                    "invalid output format, must be text, toml, or json",
+                )
+            }
+        })
+    }
+}
+
 #[derive(clap::Parser)]
 struct Opts {
     /// Elide all-zero entries from results
     #[clap(short)]
     zero_elide: bool,
 
-    /// Emit toml instead of text
-    #[clap(short)]
-    toml_output: bool,
+    /// Emit output in the specified format
```

Comment:
If clap isn't already magically printing the options for this, perhaps include them in a parenthetical?

Reply:
It's not; fixed in 41fc83815d2d.